### PR TITLE
Add generic trait to allow real cryptography

### PR DESCRIPTION
**Pull request by ma2bd**
The mega trait `SmrContext` now includes a trait `CryptographicModule` which provides hashing and signing (and the corresponding data types).

While we're at it, I also made the notion of command and state generic.

The generic types are verbose but I find it personally very satisfying to be able to write unit test with simplistic types in the simulated world.

### CODE REVIEW

File: bft-lib/src/base_types.rs [hunk 1]
```diff
@@ -1,41 +1,33 @@
...
 #[cfg(test)]
 #[path = "unit_tests/base_type_tests.rs"]
 mod base_type_tests;
 
+// TODO: add error handling (using thiserror to define an error type?) + remove Unpin
+// Alternatively, we may want to use a generic associated type when there are available on
+// rust-stable:   https://github.com/rust-lang/rust/issues/44265
+pub type AsyncResult<T> = Box<dyn std::future::Future<Output = T> + Unpin + 'static>;
+
 pub type Result<T> = std::result::Result<T, Error>;
```
Comment (asonnino) [inline bft-lib/src/base_types.rs:17]: It seems that `anyhow` already has a `Result` type defined as you did:
```
use anyhow::Result;
```

File: bft-lib/src/simulated_context.rs [hunk 2]
```diff
@@ -197,8 +237,53 @@ impl EpochReader for SimulatedContext {
...
+    }
+
+    fn author(&self) -> &Self::Author {
+        &self.config.author
+    }
+
+    fn sign(&mut self, hash: Self::HashValue) -> Result<Self::Signature> {
+        Ok(Signature(self.config.author.0, hash))
+    }
+}
```
Comment (asonnino) [inline bft-lib/src/simulated_context.rs:283]: Don't we need a batch verify here?

Comment (ma2bd) [reply bft-lib/src/simulated_context.rs:283]: Yes we should add it sometimes.

File: bft-lib/src/base_types.rs (hunk 1 above)
Comment (ma2bd) [reply bft-lib/src/base_types.rs:17]: Good point. This said, we should probably not use `anyhow::Error` for library crates.

**Pull request merged**